core/src/main/java/org/infinispan/notifications/cachelistener/filter/FilterIndexingServiceProvider.java

```diff
@@ -0,0 +1,133 @@
+package org.infinispan.notifications.cachelistener.filter;
+
+import org.infinispan.metadata.Metadata;
+import org.infinispan.notifications.cachelistener.CacheEntryListenerInvocation;
+import org.infinispan.notifications.cachelistener.event.CacheEntryEvent;
+import org.infinispan.notifications.cachelistener.event.Event;
+
+import java.lang.annotation.Annotation;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+/**
+ * A service provider for filter indexing services. This is supposed to perform the filtering operation in a more
+ * efficient way than directly executing the filter by calling the {@link org.infinispan.notifications.cachelistener.filter.CacheEventFilterConverter#filterAndConvert(Object,
+ * Object, Metadata, Object, Metadata, EventType)} method. Implementations are discovered via the {@link
+ * java.util.ServiceLoader} or {@link org.infinispan.commons.util.ServiceFinder} mechanism. Implementations may have
+ * their dependencies injected using the {@link org.infinispan.factories.annotations.Inject} annotation.
+ *
+ * @author anistor@redhat.com
+ * @since 7.2
+ */
+public interface FilterIndexingServiceProvider {
+
+   /**
+    * A wrapper around a {@link CacheEntryListenerInvocation} that keeps a reference to the {@link
+    * FilterIndexingServiceProvider} instance that handles this invocation. All methods are delegated to the wrapped
+    * invocation except {@link CacheEntryListenerInvocation#invoke(CacheEntryEvent, boolean)} and {@link
+    * CacheEntryListenerInvocation#invoke(Object)}. FilterIndexingServiceProvider implementors must extends this class
+    * and implement its abstract {@link #unregister} method.
+    *
+    * @param <K> cache key type
+    * @param <V> cache value type
+    */
+   abstract class DelegatingCacheEntryListenerInvocation<K, V> implements CacheEntryListenerInvocation<K, V> {
```

Comment:
Having an implementation class like this in the interface seems wrong to me.  Can we refactor this to a separate class?